File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/NodeRetrofitHelper.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node;

import com.flowcrypt.email.node.NodeSecret;
import com.flowcrypt.email.util.GeneralUtil;
import com.google.gson.Gson;

import java.io.IOException;
import java.security.cert.X509Certificate;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLSession;

import androidx.annotation.NonNull;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;

/**
 * This class helps to receive instance of {@link Retrofit} for requests to the Node.js server.
 *
 * @author Denis Bondarenko
 * Date: 1/13/19
 * Time: 3:28 PM
 * E-mail: DenBond7@gmail.com
 */
public final class NodeRetrofitHelper {
  private static final int TIMEOUT = 30;
  private static NodeRetrofitHelper ourInstance = new NodeRetrofitHelper();
  private OkHttpClient okHttpClient;
  private Retrofit retrofit;
  private Gson gson;

  private NodeRetrofitHelper() {
  }

  public static NodeRetrofitHelper getInstance() {
    return ourInstance;
  }

  public void init(NodeSecret nodeSecret) {
    okHttpClient = getOkHttpClientBuilder(nodeSecret).build();
    gson = NodeGson.getInstance().getGson();

    Retrofit.Builder retrofitBuilder = new Retrofit.Builder()
        .baseUrl("https://localhost:" + nodeSecret.getPort() + "/")
        .addConverterFactory(NodeConverterFactory.create(gson))
        .client(okHttpClient);

    retrofit = retrofitBuilder.build();
  }

  public OkHttpClient getOkHttpClient() {
    return okHttpClient;
  }

  public Retrofit getRetrofit() {
    return retrofit;
  }

  public Gson getGson() {
    return gson;
  }

  @NonNull
  private OkHttpClient.Builder getOkHttpClientBuilder(NodeSecret nodeSecret) {
    OkHttpClient.Builder builder = new OkHttpClient.Builder()
        .connectTimeout(TIMEOUT, TimeUnit.SECONDS)
        .readTimeout(TIMEOUT, TimeUnit.SECONDS)
        .writeTimeout(TIMEOUT, TimeUnit.SECONDS)
        .addInterceptor(headersInterceptor(nodeSecret))
        .sslSocketFactory(nodeSecret.getSslSocketFactory(), nodeSecret.getSslTrustManager())
        .followRedirects(false)
        .followSslRedirects(false)
        .hostnameVerifier(trustOurOwnCrtHostnameVerifier(nodeSecret));

    if (GeneralUtil.isDebugBuild()) {
      builder.addInterceptor(getHttpLoggingInterceptor());
    }

    return builder;
  }

  @NonNull
  private Interceptor getHttpLoggingInterceptor() {
    return new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY);
  }

  private Interceptor headersInterceptor(final NodeSecret nodeSecret) {
    return new Interceptor() {
      @Override
      public okhttp3.Response intercept(@NonNull Chain chain) throws IOException {
        okhttp3.Request request = chain.request();
        Headers headers = request
            .headers()
            .newBuilder()
            .add("Authorization", nodeSecret.getAuthHeader())
            .add("Connection", "Keep-Alive")
            .build();
        request = request.newBuilder().headers(headers).build();
        return chain.proceed(request);
      }
    };
  }

  private HostnameVerifier trustOurOwnCrtHostnameVerifier(final NodeSecret nodeSecret) {
    return new HostnameVerifier() {
      @Override
      public boolean verify(String host, SSLSession session) {
        try {
          X509Certificate crt = (X509Certificate) session.getPeerCertificates()[0];
          if (!NodeSecret.HOSTNAME.equals(host) || !NodeSecret.CRT_SUBJECT.equals(crt.getSubjectDN().getName())) {
            return false;
          }
          return crt.getSerialNumber().equals(nodeSecret.getSslCrtSerialNumber());
        } catch (Exception e) {
          return false;
        }
      }
    };
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/retrofit/node/NodeRetrofitHelper.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.retrofit.node

import android.content.Context
import androidx.annotation.WorkerThread
import androidx.preference.PreferenceManager
import com.flowcrypt.email.BuildConfig
import com.flowcrypt.email.Constants
import com.flowcrypt.email.api.retrofit.node.gson.NodeGson
import com.flowcrypt.email.node.NodeSecret
import com.flowcrypt.email.util.GeneralUtil
import com.flowcrypt.email.util.SharedPreferencesHelper
import com.google.gson.Gson
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import java.security.cert.X509Certificate
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import javax.net.ssl.HostnameVerifier

/**
 * This class helps to receive instance of [Retrofit] for requests to the Node.js server.
 *
 * @author Denis Bondarenko
 * Date: 1/13/19
 * Time: 3:28 PM
 * E-mail: DenBond7@gmail.com
 */
//todo-denbond7 We still receive ACRA reports that retrofit: Retrofit? = null, need to fix this issue
object NodeRetrofitHelper {
  private const val TIMEOUT = 90
  private var okHttpClient: OkHttpClient? = null
  private val countDownLatch: CountDownLatch = CountDownLatch(1)

  @Volatile
  private var retrofit: Retrofit? = null
  var gson: Gson = NodeGson.gson

  @WorkerThread
  fun init(context: Context, nodeSecret: NodeSecret) {
    okHttpClient = getOkHttpClientBuilder(context, nodeSecret).build()

    val retrofitBuilder = Retrofit.Builder()
      .baseUrl("https://localhost:" + nodeSecret.port + "/")
      .addConverterFactory(NodeConverterFactory.create(gson))
      .client(okHttpClient!!)

    retrofit = retrofitBuilder.build()
    countDownLatch.countDown()
  }

  @JvmStatic
  fun getRetrofit(): Retrofit? {
    if (retrofit == null) {
      countDownLatch.await(60, TimeUnit.SECONDS)
    }

    return retrofit
  }

  private fun getOkHttpClientBuilder(
    context: Context,
    nodeSecret: NodeSecret
  ): OkHttpClient.Builder {
    val builder = OkHttpClient.Builder()
      .connectTimeout(TIMEOUT.toLong(), TimeUnit.SECONDS)
      .readTimeout(TIMEOUT.toLong(), TimeUnit.SECONDS)
      .writeTimeout(TIMEOUT.toLong(), TimeUnit.SECONDS)
      .addInterceptor(headersInterceptor(nodeSecret))
      .sslSocketFactory(nodeSecret.sslSocketFactory, nodeSecret.sslTrustManager)
      .followRedirects(false)
      .followSslRedirects(false)
      .hostnameVerifier(trustOurOwnCrtHostnameVerifier(nodeSecret))

    if (GeneralUtil.isDebugBuild()) {
      val isNodeHttpLogEnabled =
        SharedPreferencesHelper.getBoolean(
          PreferenceManager.getDefaultSharedPreferences(context),
          Constants.PREF_KEY_IS_NODE_HTTP_DEBUG_ENABLED, BuildConfig.IS_NODE_HTTP_DEBUG_ENABLED
        )

      if (isNodeHttpLogEnabled) {
        val levelString = SharedPreferencesHelper.getString(
          PreferenceManager
            .getDefaultSharedPreferences(context),
          Constants.PREF_KEY_NODE_HTTP_LOG_LEVEL,
          BuildConfig.NODE_HTTP_LOG_LEVEL
        )

        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = HttpLoggingInterceptor.Level.valueOf(
          levelString
            ?: HttpLoggingInterceptor.Level.NONE.name
        )
        builder.addInterceptor(loggingInterceptor)
      }
    }

    return builder
  }

  private fun headersInterceptor(nodeSecret: NodeSecret): Interceptor {
    return Interceptor { chain ->
      var request: okhttp3.Request = chain.request()
      val headers = request
        .headers
        .newBuilder()
        .add("Authorization", nodeSecret.authHeader)
        .add("Connection", "Keep-Alive")
        .build()
      request = request.newBuilder().headers(headers).build()
      chain.proceed(request)
    }
  }

  private fun trustOurOwnCrtHostnameVerifier(nodeSecret: NodeSecret): HostnameVerifier {
    return HostnameVerifier { host, session ->
      try {
        val crt = session.peerCertificates[0] as X509Certificate
        return@HostnameVerifier if (NodeSecret.HOSTNAME != host || NodeSecret.CRT_SUBJECT != crt.subjectDN.name) {
          false
        } else crt.serialNumber == nodeSecret.sslCrtSerialNumber
      } catch (e: Exception) {
        return@HostnameVerifier false
      }
    }
  }
}
